Implementation of query_stats function

Yejashi opened the pull request:
This PR implements the query_stats function, introducing query language functionality to the statsframe. This enhancement allows end users to perform queries based on data within the statsframe.

### Summary of Changes

To achieve this, extensive modifications were necessary. Specifically, all existing stats functions now return a list of output column names. This adjustment was essential to support the caching functionality required for the query_stats function.

#### Further Explanation

When the query_stats function is called with the squash flag enabled, nodes are sometimes merged into one. This creates two primary issues:

1.  A mismatch between the expected number of nodes in the performance data and the statsframe.
2.  Invalid data within the statsframe, as some nodes have been merged and their values aggregated.

To address these issues, the statsframe needs to be recalculated after a squash operation. To facilitate this recalculation, I added a caching mechanism. This mechanism records the most recent stats operations and stores them in `statsframe_ops_cache`. These cached operations are then reapplied on a new statsframe.

By implementing this caching functionality, we ensure that the statsframe remains consistent and accurate after squash operations, enabling reliable query results.

### Important Notice

For the methods in this PR to work in the future, we must ensure a standard across all stats function such that they must return a list of output column names. This is an integral aspect of the caching mechanism.

Yejashi committed:
added implemetation of query_stats function

Yejashi committed:
made changes to query_stats to align with dialect additions

flake, black

added tests for query_stats

fixed PR issues

prog

temporary resolve PR concerns

ilumsden commented on thicket/thicket.py:
```diff
@@ -1173,6 +1173,67 @@ def query(
         return filtered_th
 
+    def query_stats(
+        self, query_obj, squash=True, update_inc_cols=True, multi_index_mode="off"
```
The `multi_index_mode` argument should not be needed for `query_stats`. As of now, the statsframe should only have a one-level row index. `multi_index_mode` is only needed when the DataFrame used in the querying has a multi-level row index.

ilumsden commented on thicket/thicket.py:
```diff
@@ -1173,6 +1173,67 @@ def query(
+
+        index_names = self.dataframe.index.names
+        dframe_copy = self.dataframe.copy()
+        index_names = self.dataframe.index.names
```
Why is this line here twice?

ilumsden commented on thicket/thicket.py:
```diff
@@ -1173,6 +1173,67 @@ def query(
+        filtered_th.dataframe = filtered_df
+
+        if squash:
+            return filtered_th.squash(update_inc_cols=update_inc_cols)
```
At this point, the new `Thicket` object will have:
* A correctly-filtered performance DataFrame
* A correctly-filtered Graph

However, you will have either an empty statsframe or the old statsframe (depending on whether `squash` clears the statsframe).

You need to add code to update the statsframe.

ilumsden commented on thicket/tests/test_query_stats.py:
```diff
@@ -0,0 +1,177 @@
+    assert all([n in filt_th_df_nodes for n in filt_nodes])
+    assert sorted(filt_th_df_profiles.unique().to_list()) == sorted(
+        th_df_profiles.unique().to_list()
+    )
```
You are not checking anything regarding the statsframe here. You need to add code to make sure the statsframe is valid too.

ilumsden commented on thicket/tests/test_query_stats.py:
```diff
@@ -0,0 +1,177 @@
+        )
+    )
+
+    new_th = th_cj.query_stats(query, multi_index_mode="all")
```
You should not have `multi_index_mode` for `query_stats`

ilumsden reviewed (changes requested):
There's one major issue right now with `query_stats` and the corresponding tests:

They don't update or check the statsframe.

Since you are querying the statsframe, I would expect the output Thicket to contain the following:
1. A statsframe containing only nodes captured by the query 
2. A graph containing only nodes captured by the query
3. A performance dataframe containing only nodes captured by the query
4. An unmodified metadata table

Your code right now produces `Thicket` objects with elements 2, 3, and 4, but it doesn't add the filtered statsframe (i.e., element 1) back to the new `Thicket`.

ilumsden commented on thicket/tests/test_query_stats.py:
```diff
@@ -0,0 +1,192 @@
+
+    assert len(sframe_nodes) == len(match)
+    assert all([n.frame in match_frames for n in sframe_nodes])
+    assert all([n.frame["name"] in match_names for n in sframe_nodes])
```
The check on line 50 (see below) should also be duplicated for statsframe:

```python
assert all([n in filt_th_df_nodes for n in filt_nodes])
```

ilumsden commented on thicket/tests/test_query_stats.py:
```diff
@@ -0,0 +1,192 @@
+    new_th = th_cj.query_stats(query)
+
+    th.stats.mean(new_th, columns=[(0, "Min time/rank")])
+    th.stats.mean(new_th, columns=[(1, "Min time/rank")])
```
Why are you calling `th.stats.mean` again? The statsframe should not be cleared, so, unless I'm missing something, there should be no need to call `mean` again

ilumsden commented on thicket/tests/test_query_stats.py:
```diff
@@ -0,0 +1,192 @@
+    queried_frames = list(sorted([n.frame for n in queried_nodes]))
+
+    th.stats.mean(new_th, columns=[(0, "Min time/rank")])
+    th.stats.mean(new_th, columns=[(1, "Min time/rank")])
```
Again, why are you calling `th.stats.mean` here?

ilumsden commented on thicket/thicket.py:
```diff
@@ -1173,6 +1173,71 @@ def query(
+
+        sframe_copy = sframe_copy[
+            sframe_copy["node"].apply(lambda x: x in query_matches)
+        ]
```
You've already done this in a more pandas-native, performant why on line 1213 (i.e., where you define `filtered_sf_df`). Remove this code and just use `filtered_sf_df`

ilumsden commented on thicket/thicket.py:
```diff
@@ -1173,6 +1173,71 @@ def query(
+            sframe_copy["node"].apply(lambda x: x in query_matches)
+        ]
+        sframe_copy.set_index(sf_index_names, inplace=True)
+        filtered_th.statsframe.dataframe = sframe_copy
```
If you're going to start with a deepcopy of the input `Thicket` object, you need to also update `filtered_th.statsframe.graph` so that it is the same as `filtered_th.graph`.

ilumsden reviewed (changes requested):
Very good, but there's still a few things that need to be addressed

Yejashi committed:
added replay functionality to mean, others in progress

added more stats functions

added more stats functions

added additional stats func support

Yejashi committed:
changed to decorator approach

flake8, black

fixed all tests

added tests for reapply function

added prelim test for cache decorator

fixed issue caused by graph copy

pr fixes

ilumsden commented on thicket/stats/calc_boxplot_statistics.py:
```diff
@@ -108,6 +117,10 @@ def calc_boxplot_statistics(thicket, columns=[], quartiles=[0.25, 0.5, 0.75], **
 
+            output_column_names.append(
+                str((boxplot_dict.keys(), next(iter(boxplot_dict.values())).keys()))
+            )
```
This is really over-complicated for getting the metric names for multi-level columns. You need to loop over the keys of the inner dictionary and add `str((idx, key))` for each iteration. So:
```python
for k in boxplot_dict[idx]:
    output_column_names.append(str((idx, k)))
```

ilumsden commented on thicket/stats/maximum.py:
```diff
@@ -43,6 +51,7 @@ def maximum(thicket, columns=None):
     else:
         df = thicket.dataframe[columns].reset_index(level=1).groupby("node").agg(max)
         for idx, column in columns:
+            output_column_names.append(str((idx, column + "_max")))
```
This applies to more than just this file, but you don't need to pass the tuple through `str()`. A tuple of strings is hashable, so it can be used as a key in the stats cache just as easily as its "stringified" version.

ilumsden commented on thicket/stats/percentiles.py:
```diff
@@ -63,8 +70,10 @@ def percentiles(thicket, columns=None, percentiles=[0.25, 0.50, 0.75]):
+            column_name_agg = ""
             for index, percentile in enumerate(percentiles):
                 column_to_append = column + "_percentiles_" + str(int(percentile * 100))
+                column_name_agg = column_name_agg + column_to_append
```
Do not do this. These are all separate columns, so there is no reason to concatenate their names together for the cache. Just do `output_column_names.append(column_to_append)`

ilumsden commented on thicket/stats/percentiles.py:
```diff
@@ -99,6 +111,7 @@ def percentiles(thicket, columns=None, percentiles=[0.25, 0.50, 0.75]):
                     idx_level,
                     "{}_percentiles_{}".format(column, str(int(percentile * 100))),
                 )
+                column_name_agg = column_name_agg + str(column_to_append)
```
See previous comments about `str()` and concatenation

ilumsden commented on thicket/stats/preference.py:
```diff
@@ -71,6 +78,10 @@ def preference(thicket, columns, comparison_func, test="ttest", *args, **kwargs)
         thicket.statsframe.dataframe[aggregated_cols + "_mean_preferred"] = pref_mean
+
+        output_column_names.append(
+            aggregated_cols + "_std_preferred" + aggregated_cols + "_mean_preferred"
```
Again, do not just concatenate these stats metric names. This produces two stats columns, so append two column names

ilumsden commented on thicket/stats/preference.py:
```diff
@@ -96,16 +107,17 @@ def preference(thicket, columns, comparison_func, test="ttest", *args, **kwargs)
         ] = pref_mean
+
+        output_column_names.append(
+            str(col_name) + "_std_preferred" + str(col_name) + "_mean_preferred"
```
Again, no concatenation

ilumsden commented on thicket/stats/stats_utils.py:
```diff
@@ -0,0 +1,13 @@
+def cache_decorator(func):
```
Please rename this. This is a really generic name. How about `cache_stats_op`?

ilumsden commented on thicket/tests/test_query_stats.py:
```diff
@@ -0,0 +1,188 @@
+    assert len(filt_nodes) == len(match)
+    assert all([n.frame in match_frames for n in filt_nodes])
+    assert all([n.frame["name"] in match_names for n in filt_nodes])
+    assert all([n in filt_th_df_nodes for n in filt_nodes])
```
The original `assert` just checks if the nodes in the graph form a subset of the nodes in the performance dataframe. The performance dataframe could contain more nodes than the graph and still pass this test.

Instead, change this to:
```python
assert sorted(filt_nodes) == sorted(filt_th_df_nodes)
```

`Node` objects are comparable (comparison is w.r.t their `_hatchet_nid` fields). As a result, you can sort both lists and then use `==` to check that the set of nodes in the Graph and performance dataframe are truly equal.

Note that you can only do this type of comparison when all the nodes come from the same `Thicket` object. There is no guarantee that this will work across `Thicket` objects (e.g., the comparisons involving `match`, `match_frames`, and `match_names` above).

ilumsden replied:
Actually, you may need to make sure that the contents of both are unique first. This won't work if there are duplicates in either list.

ilumsden commented on thicket/tests/test_query_stats.py:
```diff
@@ -0,0 +1,188 @@
+    assert all([n.frame in match_frames for n in sframe_nodes])
+    assert all([n.frame["name"] in match_names for n in sframe_nodes])
+    # Only needed if new_statsframe=False
+    assert all([n in pd.unique(filt_th_df_nodes) for n in sframe_nodes])
```
See the comment above regarding using `==` instead of `all()`. Also, this line provides an example for how to make the lists unique first.

ilumsden commented on thicket/tests/test_stats.py:
```diff
@@ -1096,3 +1096,33 @@ def test_score_hellinger(thicket_axis_columns):
+
+    assert len(th_1.statsframe_ops_cache) == 1
+    assert (
+        len(th_1.statsframe_ops_cache[list(th_1.statsframe_ops_cache.keys())[0]]) == 1
```
There's no need for two tests here. Just copy this line into `test_reapply_statsframe_operations`.

ilumsden commented on thicket/thicket.py:
```diff
@@ -396,6 +403,7 @@ def _index(thickets, from_statsframes=False, disable_tqdm=disable_tqdm):
                 metadata=thicket_parts[4],
                 profile=thicket_parts[5],
                 profile_mapping=thicket_parts[6],
+                statsframe_ops_cache={},
```
I'd recommend adding comments here and on line 418 (I'll add a comment there) saying why we reset the cache. Just a useful thing to make sure we don't forget in the future.

Yejashi replied:
No that i think about it, there might not be a reason to reset the cache when a concat with axis set to index is performed. All that really changes is the nodes and not the column.

ilumsden commented on thicket/thicket.py:
```diff
@@ -407,6 +415,7 @@ def _columns(
                 metadata_key=metadata_key,
                 disable_tqdm=disable_tqdm,
             )
+            combined_thicket.statsframe_ops_cache = {}
```
Add a comment here too

ilumsden commented on thicket/thicket.py:
```diff
@@ -1173,6 +1185,97 @@ def query(
+        filtered_th.statsframe.dataframe = filtered_sf_df
+
+        if squash:
+            filtered_th.dataframe = filtered_df
```
This needs to happen even if we don't squash. If we change the stats data, we should change the perf data.

ilumsden commented on thicket/thicket.py:
```diff
@@ -1173,6 +1185,97 @@ def query(
+        filtered_df = dframe_copy[dframe_copy["node"].isin(query_matches)]
+        filtered_df.set_index(df_index_names, inplace=True)
+
+        filtered_th = self.deepcopy()
```
Given how much we're changing, do we really need a `deepcopy` here? It might be better to just create a brand new `Thicket` object with the filtered statsframe and dataframe, the same graph, and other bookkeeping members copied.

ilumsden replied:
@pearce8 I'd appreciate your feedback here too

Yejashi replied:
@ilumsden, suppose i created a new thicket in place of deepcopy, i still need to run squash on it so doesn't that mean there will be three thickets regardless?

michaelmckinsey1 replied:
I don't see the difference of making a new Thicket as you would need to give the constructor copies of the graph and dataframe anyway.

ilumsden replied:
The graph maybe, but you don't need to give it copies of the dataframes. Passing to a constructor will be pass by reference.

I told Befikir this at lunch, but the reason to avoid `deepcopy` is because you are creating copies of the performance dataframe and (potentially) stats dataframe only to immediately replace them. There's no reason to do that.

ilumsden commented on thicket/thicket.py:
```diff
@@ -1173,6 +1185,97 @@ def query(
+
+            final_thicket.reapply_stats_operations()
+
+            return final_thicket
```
Once `filtered_th.dataframe = filtered_df` is moved outside of this conditional, we can just do all our operations on `filtered_th`. So, this return statement will become redundant.

ilumsden commented on thicket/thicket.py:
```diff
@@ -1173,6 +1185,97 @@ def query(
+                x[0],
+                tuple(sorted(x[1].items(), key=lambda y: y[0])),
+            )
+            for k, g in groupby(
```
Given that we have a `Thicket.groupby` function, it would probably be better to change the import for groupby to just `import itertools` and then change this call to `for k, g in itertools.groupby(...):`

ilumsden commented on thicket/thicket.py:
```diff
@@ -1173,6 +1185,97 @@ def query(
+                if arg[0] is None:
+                    stats_func(self, **arg[1])
+                    continue
+                elif arg[1] is None:
```
Will these ever be `None`? If not, these conditionals aren't needed. Otherwise, these conditions should probably just convert `arg[0]` to `()` and convert `arg[1]` to `{}`.

ilumsden reviewed (changes requested):
This is very close to being good to go. I've got a few more changes, but I suspect this PR will probably be good after that.

michaelmckinsey1 commented on thicket/thicket.py:
```diff
@@ -396,6 +403,7 @@ def _index(thickets, from_statsframes=False, disable_tqdm=disable_tqdm):
                 metadata=thicket_parts[4],
                 profile=thicket_parts[5],
                 profile_mapping=thicket_parts[6],
+                statsframe_ops_cache={},
```
```suggestion
```
Why is this line necessary when it will be created in the constructor?

ilumsden replied:
I'd argue this is fine. Like you said, it's not necessary, but it does convey the intent of the call more clearly than just leaving it out.

Yejashi committed:
added functools.wraps for docstrings

pr resolve

pr fixes

ilumsden commented on thicket/stats/calc_boxplot_statistics.py:
```diff
@@ -51,6 +58,8 @@ def calc_boxplot_statistics(thicket, columns=[], quartiles=[0.25, 0.5, 0.75], **
                 col + "_outliers" + q_list: [],
             }
 
+            output_column_names.append(str(boxplot_dict.keys()))
```
You should loop over `boxplot_dict.keys()`. Right now, this code is creating an iterator of strings (although you can think of it like a list) and stringifying it. Instead, you want each key to be added to `output_column_names`

ilumsden commented on thicket/stats/calc_boxplot_statistics.py:
```diff
@@ -108,6 +117,9 @@ def calc_boxplot_statistics(thicket, columns=[], quartiles=[0.25, 0.5, 0.75], **
             }
 
+            for k in boxplot_dict[idx]:
+                output_column_names.append(str((idx, k)))
```
Why are you stringifying the tuples for multi-indexed columns? Since the contents of the tuples are strings, you can use the `tuple` in the cache without stringification. Also, by stringifying, you render the return from this function unusable by users.

ilumsden replied:
Actually, I see that you have removed the stringification in some places. Given that, just treat this comment (and the other ones that refer to this) as indicators of places where you missed removing it

ilumsden commented on thicket/stats/check_normality.py:
```diff
@@ -72,6 +80,7 @@ def check_normality(thicket, columns=None):
             .agg(stats.shapiro)
         )
         for idx, column in columns:
+            output_column_names.append(str((idx, column + "_normality")))
```
See comment about stringification in `calc_boxplot_statistics.py`

ilumsden commented on thicket/stats/correlation_nodewise.py:
```diff
@@ -100,17 +109,22 @@ def correlation_nodewise(thicket, column1=None, column2=None, correlation="pears
+                column1[1] + "_vs_" + column2[1] + " " + correlation,
+            )
+            thicket.statsframe.dataframe[column_name] = correlated
+            output_column_names.append(str(column_name))
```
See comment about stringification in `calc_boxplot_statistics.py`

ilumsden commented on thicket/stats/correlation_nodewise.py:
```diff
@@ -100,17 +109,22 @@ def correlation_nodewise(thicket, column1=None, column2=None, correlation="pears
+                column1[1] + "_vs_" + column2[1] + " " + correlation,
+            )
+            thicket.statsframe.dataframe[column_name] = correlated
+            output_column_names.append(str(column_name))
```
See comment about stringification in `calc_boxplot_statistics.py`

ilumsden commented on thicket/stats/mean.py:
```diff
@@ -44,6 +52,7 @@ def mean(thicket, columns=None):
             thicket.dataframe[columns].reset_index(level=1).groupby("node").agg(np.mean)
         )
         for idx, column in columns:
+            output_column_names.append(str((idx, column + "_mean")))
```
See comment about stringification in `calc_boxplot_statistics.py`

ilumsden commented on thicket/stats/median.py:
```diff
@@ -48,6 +56,7 @@ def median(thicket, columns=None):
             .agg(np.median)
         )
         for idx, column in columns:
+            output_column_names.append(str((idx, column + "_median")))
```
See comment about stringification in `calc_boxplot_statistics.py`

ilumsden commented on thicket/stats/minimum.py:
```diff
@@ -42,6 +50,7 @@ def minimum(thicket, columns=None):
     else:
         df = thicket.dataframe[columns].reset_index(level=1).groupby("node").agg(min)
         for idx, column in columns:
+            output_column_names.append(str((idx, column + "_min")))
```
See comment about stringification in `calc_boxplot_statistics.py`

ilumsden commented on thicket/stats/preference.py:
```diff
@@ -96,16 +106,16 @@ def preference(thicket, columns, comparison_func, test="ttest", *args, **kwargs)
         ] = pref_mean
+
+        output_column_names.append(str(col_name) + "_std_preferred")
+        output_column_names.append(str(col_name) + "_mean_preferred")
```
This is similar to the comment about stringification in `calc_boxplot_statistics.py`

When you use `str(col_name)`, you are stringifying the tuple and then adding `"_std_preferred"` or `"_mean_preferred"` to it. To fix this, just copy the indexing into `thicket.statsframe.dataframe` above. So, for example:
```python
str(col_name) + "_std_preferred"
```
would become:
```python
(col_name[0], col_name[1] + "_std_preferred")
```

ilumsden commented on thicket/stats/scoring.py:
```diff
@@ -150,9 +152,12 @@ def score(thicket, columns, output_column_name, scoring_function):
     thicket.statsframe.dataframe = thicket.statsframe.dataframe.sort_index(axis=1)
 
-    return
+    output_column_names.append(str(("Scoring", stats_frame_column_name)))
```
See comment about stringification in `calc_boxplot_statistics.py`

ilumsden commented on thicket/stats/std.py:
```diff
@@ -44,6 +52,7 @@ def std(thicket, columns=None):
     else:
         df = thicket.dataframe[columns].reset_index(level=1).groupby("node").agg(np.std)
         for idx, column in columns:
+            output_column_names.append(str((idx, column + "_std")))
```
See comment about stringification in `calc_boxplot_statistics.py`

ilumsden commented on thicket/stats/variance.py:
```diff
@@ -45,6 +53,7 @@ def variance(thicket, columns=None):
     else:
         df = thicket.dataframe[columns].reset_index(level=1).groupby("node").agg(np.var)
         for idx, column in columns:
+            output_column_names.append(str((idx, column + "_var")))
```
See comment about stringification in `calc_boxplot_statistics.py`

ilumsden commented on thicket/tests/test_query_stats.py:
```diff
@@ -0,0 +1,188 @@
+    assert all([n.frame in match_frames for n in sframe_nodes])
+    assert all([n.frame["name"] in match_names for n in sframe_nodes])
+    assert all([n in pd.unique(filt_th_df_nodes) for n in sframe_nodes])
+    assert sorted(pd.unique(filt_th_df_nodes)) == sorted(pd.unique(sframe_nodes))
```
Since this is passing unit tests, there's no need to change this. But, for future reference, it's usually a good idea to make sure you aren't calling pandas functions on non-pandas types (e.g., `pd.unique(filt_th_df_nodes)`) or vice versa (e.g., `sorted(pd.unique(...))`)

ilumsden commented on thicket/thicket.py:
```diff
@@ -1173,6 +1187,96 @@ def query(
+            inc_metrics=self.inc_metrics.copy(),
+            metadata=self.metadata,
+            profile=self.profile,
+            profile_mapping=self.profile_mapping,
```
I'd copy `self.metadata`, `self.profile`, and `self.profile_mapping` here since they shouldn't be affected by the `squash`.

For `self.metadata`, you do this with `self.metadata.copy()` (with or without `deep=True`; I don't think that really matters here).

For the other two, you can copy with `copy.copy(self.profile)` or `copy.deepcopy(self.profile)`. See Thicket's `copy` and `deepcopy` methods for examples.

ilumsden reviewed (changes requested):
I suspect that the next review will be an approval for this PR.

There are just a handful of minor things to change. Most of them amount to removing stringification of multi-level column names for the cache. The remaining changes are just minor one line things.

Yejashi committed:
pr fix

ilumsden reviewed (approved):
Looks good! @pearce8 this PR is ready for your review

pearce8 commented:
@Yejashi Do we have docs/examples (tutorial?) to go with this PR?
@Yejashi @ilumsden Any further thoughts on how we could enforce that the future functions use the decorator?  Example of using this with a different decorator (e.g., profiling the functions)?

Yejashi commented:
@pearce8 , in terms of documentation, there really isn't any meaningful difference on how the stats functions are called. All that's really changed from the user end of things is that the stats functions now return a list of output columns. This is already indicated in the docstrings and but not in the thicket documentation website. A concise sentence could accomplish this task.

In terms of queries, i think we should add a small section in the query portion of the tutorials indicating that the query language can now be used on the statsframe, albeit with a slight difference.

ilumsden commented:
@pearce8 I mentioned this on Slack in more detail, but this branch shows an example of using Befikir's cache decorator with other decorators: https://github.com/TauferLab/thicket/tree/memray_decorator_example
